File: src/main/java/org/tensorflow/keras/utils/DataUtils.java
package org.tensorflow.keras.utils;

import java.io.*;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class DataUtils {

    public static String hashFile(String fpath, String algorithm) throws IOException {
        MessageDigest instance = null;

        try {
            algorithm = canonicalDigestAlgorithmNames.get(algorithm);
            instance = MessageDigest.getInstance(algorithm);
        } catch (Exception e) {
            throw new IllegalArgumentException(
                    "Hash algorithm " + algorithm + " not found. Must be 'sha256' or 'md5'");
        }

        byte[] digest = digest(Paths.get(fpath), instance);
        return toHexString(digest);
    }

    /**
     * Downloads a file from a url.
     *
     * <p>TODO: extract options .tar.gz, .zip
     *
     * @param fname    Name of the file.
     * @param origin   Original url of the file.
     * @param fileHash The expected hash string of the file after loadData.
     * @throws IOException
     */
    public static void getFile(String fname, String origin, String fileHash, String algorithm)
            throws IOException {

        File localFile = Keras.kerasPath(fname).toFile();
        File directory = localFile.getParentFile();
        if (!directory.isDirectory()) directory.mkdirs();

        if (localFile.exists()) {
            if (fileHash != null && algorithm != null) {
                String localHash = hashFile(localFile.getPath(), algorithm);
                if (localHash.equals(fileHash)) {
                    System.out.println(fname + " already exists; no need to download.");
                    return;
                } else {
                    System.out.println(fname + " exists but is corrupted. Re-downloading...");
                    return;
                }
            }
        }

        System.out.println("Downloading " + localFile + " from " + origin);
        download(origin, localFile.toString());

        if (false &&  fileHash != null && algorithm != null) {
            String calculateHash = hashFile(localFile.getPath(), algorithm);
            if (!calculateHash.equals(fileHash)) {
                System.out.println(fileHash);
                System.out.println(calculateHash);
                throw new IOException("Download failed, check origin url: " + origin);
            }
        }
    }

    public static void getFile(String fname, String origin) throws IOException {
        getFile(fname, origin, null, null);
    }

    private static void download(String url, String path) throws IOException {
        try (BufferedInputStream input = new BufferedInputStream(new URL(url).openStream());
             FileOutputStream output = new FileOutputStream(path)) {
            byte buffer[] = new byte[4096];
            for (int count; (count = input.read(buffer, 0, buffer.length)) != -1; ) {
                output.write(buffer, 0, count);
            }
        }
    }

    private static byte[] digest(Path path, MessageDigest algorithm) throws IOException {
        try (InputStream fis = Files.newInputStream(path);
             BufferedInputStream bis = new BufferedInputStream(fis);
             DigestInputStream dis = new DigestInputStream(bis, algorithm)) {
            algorithm.reset();
            while (dis.read() != 1) {
            }
            return algorithm.digest();
        }
    }

    private static String toHexString(byte[] bytes) {
        StringBuffer buffer = new StringBuffer();
        for (int i = 0; i < bytes.length; ++i) {
            byte next = bytes[i];
            if (next < 0x10) buffer.append('0');
            buffer.append(next & 0xFF);
        }

        return buffer.toString();
    }

    static Map<String, String> canonicalDigestAlgorithmNames =
            Collections.unmodifiableMap(
                    new HashMap<String, String>() {
                        {
                            put("MD5", "MD5");
                            put("md5", "MD5");
                            put("SHA-256", "SHA-256");
                            put("sha256", "SHA-256");
                        }
                    });
}

File: src/main/scala/org/tensorflow/keras/utils/DataUtils.scala
package org.tensorflow.keras.utils

import java.io.{BufferedInputStream, FileOutputStream, IOException}
import java.net.URL
import java.nio.file.{Files, Path, Paths}
import java.security.{DigestInputStream, MessageDigest}
import scala.util.control.NonFatal

object DataUtils {
  object Checksum extends Enumeration {
    val md5, sha256 = Value
  }
  type Checksum = Checksum.Value

  @throws[IOException]
  def hashFile(path: String, algorithm: Checksum): String = {
    try {
      val alg       = if (algorithm == Checksum.sha256) "SHA-256" else "MD5"
      val instance  = MessageDigest.getInstance(alg)
      val d         = digest(Paths.get(path), instance)
      toHexString(d)
    } catch {
      case NonFatal(_) =>
        throw new IllegalArgumentException(s"Hash algorithm $algorithm not found. Must be 'sha256' or 'md5'")
    }
  }

  /** Downloads a file from a url.
    *
    * <p>TODO: extract options .tar.gz, .zip
    *
    * @param fname    Name of the file.
    * @param origin   Original url of the file.
    * @param fileHash The expected hash string of the file after loadData.
    * @throws IOException
    */
  @throws[IOException]
  def getFile(fname: String, origin: String, fileHash: String, algorithm: Checksum): Unit = {
    val localFile = Keras.kerasPath(fname).toFile
    val directory = localFile.getParentFile
    if (!directory.isDirectory) directory.mkdirs()
    if (localFile.exists() && fileHash != null && algorithm != null) {
      val localHash = hashFile(localFile.getPath, algorithm)
      if (localHash == fileHash) {
        println(s"$fname already exists; no need to download.")
        return
      } else {
        println(s"$fname exists but is corrupted. Re-downloading...")
      }
    }
    println(s"Downloading $localFile from $origin")
    download(origin, localFile.toString)
    if (fileHash != null && algorithm != null) {
      val localHash = hashFile(localFile.getPath, algorithm)
      if (localHash != fileHash) {
        println(s"Expected: $fileHash")
        println(s"Found   : $localHash")
        throw new IOException(s"Download failed, check origin url: $origin")
      }
    }
  }

  @throws[IOException]
  def getFile(fname: String, origin: String): Unit =
    getFile(fname, origin, null, null)

  @throws[IOException]
  private def download(url: String, path: String): Unit = {
    val input = new BufferedInputStream(new URL(url).openStream())
    try {
      val output = new FileOutputStream(path)
      try {
        val buffer = new Array[Byte](4096)
        while ({
          val count = input.read(buffer, 0, buffer.length)
          count != -1 && {
            output.write(buffer, 0, count)
            true
          }
        }) ()

      } finally {
        output.close()
      }
    } finally {
      input.close()
    }
  }

  @throws[IOException]
  private def digest(path: Path, algorithm: MessageDigest): Array[Byte] = {
    val input = new BufferedInputStream(Files.newInputStream(path))
    try {
      val dis = new DigestInputStream(input, algorithm)
      // algorithm.reset()
      while (dis.read() != -1) ()
      dis.close()
      algorithm.digest()
    } finally {
      input.close()
    }
  }

//  private def digest(path: Path, algorithm: MessageDigest): Array[Byte] = try {
//    val fis = Files.newInputStream(path)
//    try {
//      val arr = new Array[Byte](fis.available())
//      fis.read(arr)
//      algorithm.digest(arr)
//    } finally {
//      fis.close()
//    }
//  }

  private final val HEX = "0123456789abcdef"

  private def toHexString(bytes: Array[Byte]): String = {
    val len = bytes.length
    val sb  = new StringBuffer(len << 1)
    val h   = HEX
    var i   = 0
    while (i < len) {
      val next  = bytes(i)
      val hi    = (next >> 4) & 0x0F
      val lo    =  next       & 0x0F
      sb.append(h.charAt(hi))
      sb.append(h.charAt(lo))
      i += 1
    }
    sb.toString
  }
}
